File: Java/virtual-world-simulator/src/App/GameManager.java
package App;

import Common.Point;
import Entities.Animals.Animal;
import Entities.Animals.Human;
import Entities.Entity;
import Entities.Plants.Plant;
import GUI.MapElement;

import javax.swing.*;
import java.awt.*;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.awt.event.KeyEvent;
import java.awt.event.KeyListener;
import java.io.*;
import java.util.PriorityQueue;
import java.util.Random;
import java.util.Scanner;

public class GameManager
{
    private static volatile GameManager instance;
    private JLabel menuText;
    private JButton abilityButton;
    private JTextArea logArea;
    private World world;
    private PriorityQueue<Entity> entities;
    private Human player;
    private int round;
    private int abilityDuration;
    private int abiliyCooldown;

    public static GameManager GetInstance()
    {
        GameManager result = instance;
        if(result != null)
        {
            return result;
        }
        synchronized (GameManager.class)
        {
            if(instance == null)
            {
                instance = new GameManager();
            }
            return instance;
        }
    }

    private GameManager()
    {
        world = new World();
        round = 0;
        abilityDuration = 0;
        abiliyCooldown = 5;
        entities = new PriorityQueue<Entity>(Entity::compareTo);
    }

    private void Setup(int width, int height)
    {
        world.setWorldSize(width, height);
        PlaceEntities();
        WindowManager window = WindowManager.GetInstance();
        Common.Point fixedSize = window.getFixedSize(1200, 800);
        window.ToggleWindowVisibility();
        window.setWindowSize(fixedSize.x, fixedSize.y);
        window.setWindowTitle("Adrian Ściepura 193350");

        SetupGame();
        SetupMenu();

        JFrame frame = window.getWindow();
        frame.addKeyListener(new KeyListener()
        {
            @Override
            public void keyTyped(KeyEvent e)
            {

            }

            @Override
            public void keyPressed(KeyEvent e)
            {
                int keyCode = e.getKeyCode();

                if(!world.CheckIfGameOver() && player.setNewPosition(keyCode))
                {
                    round++;
                    if(player.CheckIfAbilityTurnedOn())
                    {
                        abilityDuration--;
                        if(abilityDuration == 0)
                            player.ToggleAbility(false);
                    }
                    else
                    {
                        if(abiliyCooldown > 0)
                            abiliyCooldown--;
                    }

                    for(int i = 0; i < world.worldHeight; i++)
                        for(int j = 0; j < world.worldWidth; j++)
                        {
                            Entity entity = world.map[i][j].getEntity();
                            if(entity != null)
                                entities.add(entity);
                        }

                    world.ClearLogs();
                    Entity entity = null;
                    while(!entities.isEmpty())
                    {
                        entity = entities.poll();
                        if(entity.checkIfAlive())
                            entity.Update();
                    }
                }

                UpdateMenu();
            }

            @Override
            public void keyReleased(KeyEvent e)
            {

            }
        });
    }

    private void SetupMenu()
    {
        WindowManager window = WindowManager.GetInstance();
        JPanel menuPanel = window.getMenuPanel();
        JPanel gamePanel = window.getGamePanel();
        menuText = new JLabel();
        Common.Point fixedSize = window.getFixedSize(menuText.getFont().getSize()*2, 0);
        menuText.setFont(new Font(menuText.getFont().getName(), Font.PLAIN, fixedSize.x));
        menuText.setHorizontalAlignment(SwingConstants.CENTER);
        menuText.setHorizontalTextPosition(SwingConstants.CENTER);
        menuText.setAlignmentX(Component.CENTER_ALIGNMENT);
        menuPanel.add(menuText);

        JButton changeAppearanceButton = new JButton("Change appearance");
        changeAppearanceButton.setFont(new Font(menuText.getFont().getName(), Font.PLAIN, fixedSize.x));
        changeAppearanceButton.setFocusable(false);
        changeAppearanceButton.setAlignmentX(Component.CENTER_ALIGNMENT);
        changeAppearanceButton.addActionListener(e -> {
            MapElement.ChangeAppearance();
            gamePanel.repaint();
        });
        menuPanel.add(changeAppearanceButton);

        abilityButton = new JButton("Activate ability");
        abilityButton.setFont(new Font(menuText.getFont().getName(), Font.PLAIN, fixedSize.x));
        abilityButton.setFocusable(false);
        abilityButton.setAlignmentX(Component.CENTER_ALIGNMENT);
        abilityButton.addActionListener(new ActionListener()
        {
            @Override
            public void actionPerformed(ActionEvent e)
            {
                player.ToggleAbility(true);
                abilityButton.setEnabled(false);
                abilityDuration = 5;
                abiliyCooldown = 5;
                UpdateMenu();
                System.out.println("Activated ability");
            }
        });
        menuPanel.add(abilityButton);

        JButton saveButton = new JButton("Save game");
        saveButton.setFont(new Font(menuText.getFont().getName(), Font.PLAIN, fixedSize.x));
        saveButton.setFocusable(false);
        saveButton.setAlignmentX(Component.CENTER_ALIGNMENT);
        saveButton.addActionListener(e -> {
            SaveGame();
        });
        menuPanel.add(saveButton);

        JButton loadButton = new JButton("Load game");
        loadButton.setFont(new Font(menuText.getFont().getName(), Font.PLAIN, fixedSize.x));
        loadButton.setFocusable(false);
        loadButton.setAlignmentX(Component.CENTER_ALIGNMENT);
        loadButton.addActionListener(e -> {
            LoadGame();
        });
        menuPanel.add(loadButton);

        JButton newGameButton = new JButton("New game");
        newGameButton.setFont(new Font(menuText.getFont().getName(), Font.PLAIN, fixedSize.x));
        newGameButton.setFocusable(false);
        newGameButton.setAlignmentX(Component.CENTER_ALIGNMENT);
        newGameButton.addActionListener(e -> {
            NewGame();
        });
        menuPanel.add(newGameButton);

        JPanel panel = new JPanel();
        panel.setBackground(menuPanel.getBackground());
        logArea = new JTextArea(20, 25);
        logArea.setEditable(false);
        logArea.setFocusable(false);
        JScrollPane scrollPane = new JScrollPane(logArea, JScrollPane.VERTICAL_SCROLLBAR_ALWAYS, JScrollPane.HORIZONTAL_SCROLLBAR_NEVER);
        panel.add(scrollPane);
        panel.setPreferredSize(new Dimension(100, 100));
        menuPanel.add(panel);

        UpdateMenu();
        menuPanel.repaint();
    }

    private void SetupGame()
    {
        WindowManager window = WindowManager.GetInstance();
        JPanel gamePanel = window.getGamePanel();

        if(gamePanel.getComponents().length > 0)
            gamePanel.removeAll();

        gamePanel.setLayout(new GridLayout(world.worldHeight, world.worldWidth));
        for(int i = 0; i < world.worldHeight; i++)
            for(int j = 0; j < world.worldWidth; j++)
                gamePanel.add(world.map[i][j]);
    }

    public void Start()
    {
        Setup(20, 20);
    }

    private void PlaceEntities()
    {
        char[] types = { 'A', 'F', 'T', 'W', 'S', 'G', 'D', 'U', 'N', 'P'};
        int freeSpace = world.worldHeight * world.worldWidth;
        int x = 0;
        int y = 0;
        int howMany = freeSpace / 3;
        int type = 0;

        player = new Human(world, new Point(1, 1));
        world.map[1][1].setEntity(player);

        Random rnd = new Random();
        for(int i = 0; i < howMany; i++)
        {
            x = rnd.nextInt(world.worldWidth);
            y = rnd.nextInt(world.worldHeight);
            type = rnd.nextInt(10);

            if(world.map[y][x].getEntity() == null)
                world.map[y][x].setEntity(Entity.getEntityFromSymbol(this.world, new Point(y, x), types[type]));
        }
    }

    private void UpdateMenu()
    {
        menuText.setText(
                        "<html><br><p style=\"text-align: center;\"><b>Adrian Ściepura</b>" +
                        "<br>Number of Animals: <b>" + Animal.getNumberOfAnimals() +
                        "</b><br>Number of Plants: <b>" + Plant.getNumberOfPlants() +
                        "</b><br>Round: <b>" + round +
                        "</b><br>Ability Cooldown: <b>" + abiliyCooldown + "</b></p><br></html>");

        if(abiliyCooldown == 0)
            abilityButton.setEnabled(true);
        else
            abilityButton.setEnabled(false);

        logArea.setText(world.getLogs());
    }

    private void SaveGame()
    {
        try
        {
            FileWriter saveFile = new FileWriter("save.txt");
            saveFile.write(world.worldHeight + " " + world.worldWidth + " " + round + " " + world.CheckIfGameOver() + "\n");
            saveFile.write(abilityDuration + " " + abiliyCooldown + "\n");
            for(int i = 0; i < world.worldHeight; i++)
            {
                for(int j = 0; j < world.worldWidth; j++)
                {
                    Entity entity = world.map[i][j].getEntity();
                    if(entity != null)
                        saveFile.write(entity.getSymbol() + " " + entity.getPosition().x + " " + entity.getPosition().y + " " + entity.getLifespan() + " " + entity.getStrength() + "\n");
                }
            }
            saveFile.close();

        } catch (IOException e)
        {
            System.out.println("An error has occured");
            e.printStackTrace();
        }

    }

    private void LoadGame()
    {
        try
        {
            FileReader saveFile = new FileReader("save.txt");
            Scanner sc = new Scanner(saveFile);

            int height = sc.nextInt();
            int width = sc.nextInt();
            round = sc.nextInt();
            boolean gameOver = sc.nextBoolean();
            abilityDuration = sc.nextInt();
            abiliyCooldown = sc.nextInt();

            world.Reset();
            world.ClearLogs();
            world.setWorldSize(width, height);
            if(gameOver)
                world.EndGame();
            SetupGame();
            while(sc.hasNextLine())
            {
                char symbol = sc.next().charAt(0);
                int x = sc.nextInt();
                int y = sc.nextInt();
                int lifespan = sc.nextInt();
                int strength = sc.nextInt();
                sc.nextLine();

                Entity entity = Entity.getEntityFromSymbol(world, new Point(x, y), symbol);
                entity.setLifespan(lifespan);
                entity.setStrength(strength);
                world.map[x][y].setEntity(entity);

                if(symbol == 'H')
                    player = (Human) entity;
            }

            saveFile.close();
            sc.close();

            UpdateMenu();
        }
        catch (FileNotFoundException e)
        {
            System.out.println("Save file not found");
        }
        catch (IOException e)
        {
            throw new RuntimeException(e);
        }

    }

    private void NewGame()
    {
        world.Reset();
        round = 0;
        abiliyCooldown = 5;
        abilityDuration = 0;
        PlaceEntities();
        world.ClearLogs();
        UpdateMenu();
    }

    public World getWorld()
    {
        return world;
    }
}

File: Python/virtual-world-simulator/src/App/GameManager.py
import os.path
import random
import queue
import pygame
from src.App.World import World
from src.Entity.Animals.Human import Human
from src.UI.Button import Button
from src.App.Other import getEntityFromSymbol


class GameManager:
    BACKGROUND_COLOR = (124, 202, 146)

    def __init__(self):
        pygame.init()
        self._screen = pygame.display.set_mode((1200, 700))
        pygame.display.set_caption('Adrian Ściepura 193350')
        self._newGameButton = Button(self._screen, 800, 200, 200, 50, "New Game", self.newGame)
        self._saveGameButton = Button(self._screen, 800, 260, 200, 50, "Save Game", self.saveGame)
        self._loadGameButton = Button(self._screen, 800, 320, 200, 50, "Load Game", self.loadGame)
        self._abilityButton = Button(self._screen, 800, 380, 200, 50, "Activate ability", self.toggleAbility)
        self._abilityButton.enabled = False
        self._buttons = [self._newGameButton, self._saveGameButton, self._loadGameButton, self._abilityButton]
        self._world = World(self._screen, 20, 20)
        self._entities = queue.PriorityQueue()
        self._player = None
        self._round = 0
        self._abilityCooldown = 5
        self._abilityDuration = 0
        self.placeEntities()

        if not os.path.exists("save.txt"):
            self._loadGameButton.enabled = False

    def start(self):
        while True:
            for event in pygame.event.get():
                if event.type == pygame.QUIT:
                    pygame.quit()
                    exit()
                elif event.type == pygame.KEYUP:
                    if self._player.setNewPosition(event.key) and not self._world.gameOver:
                        self.update()

                self._world.handleEvent(event)
                for bt in self._buttons:
                    bt.handleEvent(event)

            self._screen.fill(self.BACKGROUND_COLOR)
            self._world.drawWorld()
            self.displayMenu()
            pygame.display.update()

    def update(self):
        self._round += 1
        self._world.clearLogs()
        if self._player.abilityTurnedOn:
            if self._abilityDuration == 0:
                self._player.abilityTurnedOn = False
            self._abilityDuration -= 1
        elif self._abilityCooldown > 0:
            self._abilityCooldown -= 1
            if self._abilityCooldown == 0:
                self._abilityButton.enabled = True

        for i in range(self._world.height):
            for j in range(self._world.width):
                element = self._world.getMapElement(i, j)
                if element is not None:
                    self._entities.put(element)

        while not self._entities.empty():
            entity = self._entities.get()
            if entity.checkIfAlive():
                entity.update()

    def placeEntities(self):
        self._player = Human(self._world, (3, 3))
        self._world.setMapElement(3, 3, self._player)

        freeSpace = self._world.width * self._world.height
        availableTypes = ['W', 'A', 'F', 'S', 'T', 'C', 'G', 'D', 'U', 'N', 'P']
        x = 0
        y = 0
        typeToPlace = 0
        howMany = int(freeSpace / 3)
        for i in range(howMany):
            y = random.randint(0, self._world.height - 1)
            x = random.randint(0, self._world.width - 1)
            typeToPlace = random.randint(0, 10)

            if self._world.getMapElement(y, x) is None:
                self._world.setMapElement(y, x, getEntityFromSymbol(self._world, (y, x), availableTypes[typeToPlace]))

    def displayMenu(self):
        menuText = [("Adrian Ściepura", 50), ("Round: " + str(self._round), 35),
                    ("Ability Cooldown: " + str(self._abilityCooldown) if not self._player.abilityTurnedOn else "Ability Duration: " + str(self._abilityDuration), 35)]
        y = 50
        for element in menuText:
            font = pygame.font.Font(None, element[1])
            self._screen.blit(font.render(element[0], False, 'White'), (800, y))
            y += element[1]

        for bt in self._buttons:
            bt.draw()

        if self._world.gameOver:
            font = pygame.font.Font(None, 60)
            self._screen.blit(font.render("Game Over", False, 'Red'), (250, 300))

        self._world.showLastLogs((800, 450), 10)

    def newGame(self):
        self._world.restart()
        self.placeEntities()
        self._round = 0
        self._abilityCooldown = 5
        self._abilityDuration = 0

    def saveGame(self):
        file = open("save.txt", "w")
        file.write(str(self._world.width) + ' ' + str(self._world.height) + ' ' + str(self._round) + ' ' + str(self._world.gameOver) + '\n')
        file.write(str(self._abilityDuration) + ' ' + str(self._abilityCooldown) + '\n')
        for i in range(self._world.height):
            for j in range(self._world.width):
                entity = self._world.getMapElement(i, j)
                if entity is not None:
                    file.write(
                        str(entity.symbol) + ' ' + str(entity.position[0]) + ' ' + str(entity.position[1]) + ' ' + str(
                            entity.lifespan) + ' ' + str(entity.strength) + '\n')
        file.close()
        self._loadGameButton.enabled = True

    def loadGame(self):
        if not os.path.exists("save.txt"):
            return

        file = open("save.txt", "r")
        self._world.restart()
        line = file.readline().strip().split()
        self._world.width = int(line[0])
        self._world.height = int(line[1])
        self._round = int(line[2])
        self._world.gameOver = True if line[3] == "True" else False

        line = file.readline().strip().split()
        self._abilityDuration = int(line[0])
        self._abilityCooldown = int(line[1])


        for line in file:
            letter, x, y, lifespan, strength = line.strip().split()
            x = int(x)
            y = int(y)
            lifespan = int(lifespan)
            strength = int(strength)
            entity = getEntityFromSymbol(self._world, (x, y), letter)
            entity.lifespan = lifespan
            entity.strength = strength
            self._world.setMapElement(x, y, entity)
            if letter == 'H':
                self._player = entity

        if self._abilityDuration > 0:
            self._abilityButton.enabled = False
            self._player.abilityTurnedOn = True

        file.close()

    def toggleAbility(self):
        self._player.abilityTurnedOn = True
        self._abilityButton.enabled = False
        self._abilityDuration = 5
        self._abilityCooldown = 5
